File: src/lib/batch-enrich.mjs
// Batch enrichment script — processes all cached animals that lack LLM data,
// with a 4-second delay between each to avoid API rate limits (429).
//
// Usage: node src/lib/batch-enrich.mjs

import ZAI from "z-ai-web-dev-sdk";
import { readFileSync, writeFileSync, readdirSync, existsSync } from "fs";
import path from "path";
import { setTimeout as sleep } from "timers/promises";

const ROOT = process.cwd();
const CACHE_DIR = path.join(ROOT, "data", "cache", "wikipedia");

const EXTRACTION_PROMPT = `You are a precise wildlife data extractor. Given the Wikipedia article text about an animal, extract structured facts. Only include values actually stated in the text — leave a field empty ("") if not mentioned. Use metric units (m, cm, kg, km/h, years).

Output STRICT JSON (no markdown fences) with this exact shape:
{
  "scientific_name": "Panthera tigris",
  "young_name": "cub",
  "group_name": "pride",
  "physical": { "length": "", "height": "", "weight": "", "top_speed": "", "lifespan": "" },
  "ecology": { "diet": "", "habitat": "", "locations": "", "group_behavior": "", "conservation_status": "", "biggest_threat": "", "distinctive_features": [] },
  "classification": { "kingdom": "", "phylum": "", "class": "", "order": "", "family": "", "genus": "", "species": "" },
  "reproduction": { "gestation_period": "", "average_litter_size": "", "name_of_young": "" },
  "time_period_text": ""
}

Rules:
- diet: one of Carnivore, Herbivore, Omnivore, Piscivore, Insectivore, Nectarivore, Scavenger, Filter feeder
- conservation_status: one of Least Concern, Near Threatened, Vulnerable, Endangered, Critically Endangered, Extinct in the Wild, Extinct, Data Deficient
- Use ranges (en-dash –) when given. Single value otherwise.
- For egg-layers: gestation=incubation, litter=clutch size.
- classification: fill all 7 ranks if mentioned. Blank if not.
- Do NOT invent values. Empty "" is better than a guess.`;

function mergeExtracted(animal, extracted) {
  const merged = { ...animal };
  if (extracted.scientific_name?.trim()) merged.scientific_name = extracted.scientific_name.trim();
  if (extracted.young_name?.trim()) merged.young_name = extracted.young_name.trim();
  if (extracted.group_name?.trim()) merged.group_name = extracted.group_name.trim();
  if (extracted.physical) {
    merged.physical = { ...(animal.physical || {}) };
    for (const [k, v] of Object.entries(extracted.physical)) {
      if (typeof v === "string" && v.trim()) merged.physical[k] = v;
    }
  }
  if (extracted.ecology) {
    merged.ecology = { ...(animal.ecology || {}) };
    for (const [k, v] of Object.entries(extracted.ecology)) {
      if (typeof v === "string" && v.trim()) merged.ecology[k] = v;
      else if (Array.isArray(v) && v.length) merged.ecology[k] = v;
    }
  }
  if (extracted.classification) {
    merged.classification = { ...(animal.classification || {}) };
    for (const [k, v] of Object.entries(extracted.classification)) {
      if (typeof v === "string" && v.trim()) merged.classification[k] = v;
    }
  }
  if (extracted.reproduction) {
    merged.reproduction = { ...(animal.reproduction || {}) };
    for (const [k, v] of Object.entries(extracted.reproduction)) {
      if (typeof v === "string" && v.trim()) merged.reproduction[k] = v;
    }
  }
  if (extracted.time_period_text?.trim()) {
    merged.time_period = { text: extracted.time_period_text.trim(), width: "50%", start: "Ancient", end: "Present" };
  }
  return merged;
}

async function main() {
  const zai = await ZAI.create();
  const files = readdirSync(CACHE_DIR).filter(f => f.endsWith(".json")).sort();

  let processed = 0;
  let skipped = 0;
  let failed = 0;

  for (const file of files) {
    const cacheFile = path.join(CACHE_DIR, file);
    const animal = JSON.parse(readFileSync(cacheFile, "utf-8"));

    // Skip if already has full LLM data
    if (animal.scientific_name?.trim() && (animal.classification?.class || "")?.trim()) {
      skipped++;
      continue;
    }

    // Build the text blob from article sections
    const blob = [
      animal.summary,
      animal.article?.overview,
      animal.article?.description,
      animal.article?.habitat,
      animal.article?.behavior,
      animal.article?.conservation,
    ].filter(Boolean).join("\n\n");

    if (!blob.trim()) {
      console.log(`  ✗ ${animal.name}: no text to extract from`);
      failed++;
      continue;
    }

    try {
      const response = await zai.chat.completions.create({
        messages: [{
          role: "user",
          content: `${EXTRACTION_PROMPT}\n\nAnimal name: ${animal.name}\n\nWikipedia article text:\n${blob.slice(0, 5000)}`,
        }],
        thinking: { type: "disabled" },
      });

      const content = response.choices?.[0]?.message?.content ?? "";
      const cleaned = content.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/i, "").trim();

      let extracted;
      try {
        extracted = JSON.parse(cleaned);
      } catch {
        const m = cleaned.match(/\{[\s\S]*\}/);
        extracted = m ? JSON.parse(m[0]) : {};
      }

      const enriched = mergeExtracted(animal, extracted);
      writeFileSync(cacheFile, JSON.stringify(enriched, null, 2));
      processed++;
      console.log(`  ✓ [${processed}] ${animal.name}: sci=${enriched.scientific_name || "?"} class=${enriched.classification?.class || "?"}`);
    } catch (err) {
      console.error(`  ✗ ${animal.name}: ${err.message?.slice(0, 80)}`);
      failed++;
      // If rate-limited, wait longer
      if (err.message?.includes("429")) {
        console.log("    Rate limited, waiting 30s...");
        await sleep(30000);
      }
    }

    // 4-second delay between calls to avoid rate limits
    await sleep(4000);
  }

  console.log(`\nDone. Processed: ${processed} | Skipped (already enriched): ${skipped} | Failed: ${failed}`);
}

main().catch(console.error);

File: scripts/batch_enrich.py
#!/usr/bin/env python3
"""Batch-enrich cached animals using the z-ai CLI (one process per animal).
Rate-limit-safe: 3s delay between calls."""
import json, os, re, subprocess, time, sys

CACHE_DIR = "data/cache/wikipedia"
PROMPT = """You are a precise wildlife data extractor. Given the Wikipedia article text about an animal, extract structured facts. Only include values actually stated in the text. Use metric units.

Output STRICT JSON (no markdown fences) with this shape:
{"scientific_name":"","young_name":"","group_name":"","physical":{"length":"","height":"","weight":"","top_speed":"","lifespan":""},"ecology":{"diet":"","habitat":"","locations":"","group_behavior":"","conservation_status":"","biggest_threat":"","distinctive_features":[]},"classification":{"kingdom":"","phylum":"","class":"","order":"","family":"","genus":"","species":""},"reproduction":{"gestation_period":"","average_litter_size":"","name_of_young":""},"time_period_text":""}

diet: one of Carnivore/Herbivore/Omnivore/Piscivore/Insectivore/Nectarivore/Scavenger/Filter feeder. conservation_status: one of Least Concern/Near Threatened/Vulnerable/Endangered/Critically Endangered/Extinct in the Wild/Extinct/Data Deficient. Use ranges with en-dash. For egg-layers: gestation=incubation, litter=clutch. Do NOT invent values.

Animal name: {name}

Wikipedia article text:
{text}"""

def merge(animal, extracted):
    if not extracted:
        return animal
    if extracted.get("scientific_name", "").strip():
        animal["scientific_name"] = extracted["scientific_name"].strip()
    if extracted.get("young_name", "").strip():
        animal["young_name"] = extracted["young_name"].strip()
    if extracted.get("group_name", "").strip():
        animal["group_name"] = extracted["group_name"].strip()
    for section in ["physical", "ecology", "classification", "reproduction"]:
        if extracted.get(section):
            animal[section] = {**(animal.get(section) or {})}
            for k, v in extracted[section].items():
                if isinstance(v, str) and v.strip():
                    animal[section][k] = v
                elif isinstance(v, list) and v:
                    animal[section][k] = v
    if extracted.get("time_period_text", "").strip():
        animal["time_period"] = {
            "text": extracted["time_period_text"].strip(),
            "width": "50%", "start": "Ancient", "end": "Present"
        }
    return animal

def enrich_one(cache_file):
    animal = json.load(open(cache_file))
    if animal.get("scientific_name", "").strip() and (animal.get("classification") or {}).get("class", "").strip():
        return "skip"
    art = animal.get("article") or {}
    parts = [animal.get("summary",""), art.get("overview",""), art.get("description",""),
             art.get("habitat",""), art.get("behavior",""), art.get("conservation","")]
    text = "\n\n".join(p for p in parts if p)[:5000]
    if not text.strip():
        return "no-text"
    prompt = PROMPT.replace("{name}", animal["name"]).replace("{text}", text)
    try:
        result = subprocess.run(["z-ai", "chat", "-p", prompt], capture_output=True, text=True, timeout=30)
        m = re.search(r'\{[\s\S]*"choices"[\s\S]*\}', result.stdout)
        if not m:
            return "no-json"
        resp = json.loads(m.group(0))
        content = resp["choices"][0]["message"]["content"]
        content = re.sub(r'^```(?:json)?\s*', '', content.strip())
        content = re.sub(r'\s*```$', '', content)
        try:
            extracted = json.loads(content)
        except:
            m2 = re.search(r'\{[\s\S]*\}', content)
            extracted = json.loads(m2.group(0)) if m2 else {}
        enriched = merge(animal, extracted)
        json.dump(enriched, open(cache_file, "w"), indent=2, ensure_ascii=False)
        return f"ok: sci={enriched.get('scientific_name','?')} class={(enriched.get('classification') or {}).get('class','?')}"
    except Exception as e:
        return f"error: {str(e)[:60]}"

files = sorted(f for f in os.listdir(CACHE_DIR) if f.endswith(".json"))
total = len(files)
processed = ok = skipped = failed = 0

for i, fn in enumerate(files):
    cache_file = os.path.join(CACHE_DIR, fn)
    result = enrich_one(cache_file)
    processed += 1
    if result == "skip":
        skipped += 1
        continue
    elif result.startswith("ok"):
        ok += 1
        print(f"[{processed}/{total}] OK {result}", flush=True)
    else:
        failed += 1
        print(f"[{processed}/{total}] FAIL {fn}: {result}", flush=True)
    time.sleep(3)

print(f"\nDone: {ok} enriched, {skipped} skipped, {failed} failed")
